### PR TITLE
[GEOT-7054] Like Query Addition to Geopackage

### PR DESCRIPTION
**Pull request by ydeliorman**
[GEOT-7054](https://osgeo-org.atlassian.net/browse/GEOT-7054) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

Like query was not supported in geopackage. With this PR requests using like query will be supported in geopackage.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPkgFilterToSQL.java [hunk 1]
```diff
@@ -237,4 +238,11 @@ protected Object visitBinarySpatialOperator(
 
         return extraData;
     }
+
+    @Override
+    protected FilterCapabilities createFilterCapabilities() {
```
Comment (aaime) [inline modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPkgFilterToSQL.java:243]: This is worrysome, no clue as to why FilterToSQL does not claim support for PropertyIsLike, looks like an issue that might be affecting other databases.. I've done a quick check, PostGIS and Oracle setup the filter capabilities in a different way and end up having PropertyIsLike support, but for example MySQL is affected by this problem as well.

@ydeliorman could you try declaring support for PropertyIsLike in the base class instead, and see if any test breaks?

Comment (ydeliorman) [reply modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPkgFilterToSQL.java:243]: @aaime after moving the logic should I be running the tests on the GeoTools side? Do I need to get a full run at GeoServer sider too?

Comment (aaime) [reply modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPkgFilterToSQL.java:243]: Running all database tests on your machine is quite a bit of setup, but the PR checks do it for you so... just make the change and push it here, then see what the various database test runs tell you, e.g.:

image

File: modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgFilterToSQLTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        PropertyIsLike like =
+                ff.like(
+                        ff.function("strToLowerCase", ff.property("testString")),
+                        "a_literal",
+                        "%",
+                        "-",
+                        "\\",
+                        true);
+
+        filterToSql.encode(like);
```
Comment (aaime) [inline modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgFilterToSQLTest.java:49]: So before the change in FilterToSQL this line would have thrown a "Filter type not supported" exception right?

**Review by aaime — CHANGES_REQUESTED** [review]

File: modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgFilterToSQLTest.java (hunk 2 above)
Comment (ydeliorman) [reply modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgFilterToSQLTest.java:65]: Yes, it was throwing that exception.

File: modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgFilterToSQLTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package org.geotools.geopkg;
```
[the remark this answers is not in the document]
Comment (ydeliorman) [reply modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgFilterToSQLTest.java:17]: added

**Comment by ydeliorman** [comment]
@aaime changes are done.

File: modules/library/jdbc/src/main/java/org/geotools/data/jdbc/FilterToSQL.java [hunk 4]
```diff
@@ -393,6 +393,7 @@ protected FilterCapabilities createFilterCapabilities() {
         capabilities.addAll(FilterCapabilities.SIMPLE_COMPARISONS_OPENGIS);
         capabilities.addType(PropertyIsNull.class);
         capabilities.addType(PropertyIsBetween.class);
+        capabilities.addType(PropertyIsLike.class);
```
Comment (jodygarnett) [inline modules/library/jdbc/src/main/java/org/geotools/data/jdbc/FilterToSQL.java:396]: Not sure I get it, was this filter not encoded for any of the JDBCDataStore implementations previously?

Making the change here should require testing in all the datastore’s that use FilterToSQL …

Comment (jodygarnett) [reply modules/library/jdbc/src/main/java/org/geotools/data/jdbc/FilterToSQL.java:396]: Reading previous comment it appears that PropertyIsLike was not being encoded for any of the JDBC datastore instances?

This should be a really good improvement then …

Comment (aaime) [reply modules/library/jdbc/src/main/java/org/geotools/data/jdbc/FilterToSQL.java:396]: Wrong, it was encoded in most, but not all. Some ignore the setup of FilterToSQL and do their own, which includes PropertyIsLike, but GeoPackage and MySQL inherited the base config from FilterToSQL instead.

Comment (aaime) [reply modules/library/jdbc/src/main/java/org/geotools/data/jdbc/FilterToSQL.java:396]: And yep, we have tests checking the result of PropertyIsLike, just not the encoding of it.

**Review by aaime — APPROVED** [review]

**Comment by aaime** [comment]
Thankfully we have a good battery of DBMS tests here in the PR checks covering most databases... We should probably require a database to have a PR check in order to be in supported, since changes to the code DB machinery affect them all, but nobody can run all of the database types.

**Comment by aaime** [comment]
@ydeliorman squash merged, backport as you see fit.

**Pull request merged**